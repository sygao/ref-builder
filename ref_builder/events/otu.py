from pydantic import ConfigDict, UUID4, field_serializer

from ref_builder.events.base import ApplicableEvent, EventData, Event, OTUQuery
from ref_builder.resources.otu import RepoOTU
from ref_builder.models import Molecule
from ref_builder.plan import Plan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: Plan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery

    def apply(self) -> RepoOTU:
        return RepoOTU(
            id=self.query.otu_id,
            acronym=self.data.acronym,
            excluded_accessions=set(),
            isolates=[],
            legacy_id=self.data.legacy_id,
            molecule=self.data.molecule,
            name=self.data.name,
            representative_isolate=None,
            plan=self.data.plan,
            taxid=self.data.taxid,
        )


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: Plan


class CreatePlan(ApplicableEvent):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Apply changed plan to OTU and return."""
        otu.plan = self.data.plan

        return otu


class SetRepresentativeIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetRepresentativeIsolate(ApplicableEvent):
    """An event that sets the representative isolate for an OTU."""

    data: SetRepresentativeIsolateData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Update the OTU's representative isolate and return."""
        otu.representative_isolate = self.data.isolate_id

        return otu


class UpdateExcludedAccessionsData(EventData):
    """The data for an UpdateAllowedAccessions event."""

    model_config = ConfigDict(use_enum_values=True)

    accessions: set[str]
    action: ExcludedAccessionAction

    @field_serializer("accessions")
    def serialize_accessions(self, accessions: set[str]) -> list[str]:
        return sorted(accessions)


class UpdateExcludedAccessions(ApplicableEvent):
    """An event that changes the OTU excluded accessions collection.

    This event is emitted when Genbank accessions are either
    allowed or disallowed from inclusion in the reference.
    """

    data: UpdateExcludedAccessionsData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Add accession allowance changes to OTU and return."""

        if self.data.action == ExcludedAccessionAction.ALLOW:
            for accession in self.data.accessions:
                otu.excluded_accessions.discard(accession)

        else:
            for accession in self.data.accessions:
                otu.excluded_accessions.add(accession)

        return otu
