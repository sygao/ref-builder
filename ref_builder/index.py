"""An index for improving repository performance."""

import binascii
import sqlite3
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from uuid import UUID

import orjson

from ref_builder.models import OTUMinimal
from ref_builder.resources.otu import RepoOTU


@dataclass
class EventIndexItem:
    """The event IDs associated with an OTU."""

    at_event: int
    """The latest event ID when this cache index was last updated"""

    event_ids: list[int]
    """A list of event IDs"""

    otu_id: UUID
    """The OTU ID this cache index is associated with."""


@dataclass
class Snapshot:
    """A snapshot of an OTU at a specific event."""

    at_event: int
    """The event at which the snapshot was taken."""

    otu: RepoOTU
    """The OTU that was snapshotted."""


class Index:
    """An index for rapid access to repository data.

    1. Get OTU IDs by legacy ID, name, or taxonomy ID.
    2. Save and load complete snapshots of OTUs.
    3. Get the event IDS associated with an OTU.

    """

    def __init__(self, path: Path) -> None:
        """Initialize the index."""
        self.path = path
        """The path to the index file."""

        self.path.parent.mkdir(exist_ok=True, parents=True)

        self.con = sqlite3.connect(path, isolation_level=None)
        self.con.execute("PRAGMA journal_mode = WAL")
        self.con.execute("PRAGMA synchronous = NORMAL")

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS at_event (
                id INTEGER PRIMARY KEY,
                at_event INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS events (
                at_event INTEGER,
                event_id INTEGER PRIMARY KEY,
                otu_id TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otus (
                id TEXT PRIMARY KEY,
                acronym TEXT,
                at_event INTEGER,
                legacy_id TEXT,
                name TEXT,
                otu JSONB,
                taxid INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS sequences (
                id TEXT PRIMARY KEY,
                crc TEXT,
                otu_id TEXT,
                sequence TEXT
            );
            """,
        )

        for table, column in [
            ("events", "otu_id"),
            ("otus", "id"),
            ("otus", "legacy_id"),
            ("otus", "name"),
            ("otus", "taxid"),
            ("sequences", "otu_id"),
            ("sequences", "crc"),
        ]:
            self.con.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {table}_{column} ON {table} ({column});
                """,
            )

        self.con.commit()

    @property
    def otu_ids(self) -> set[UUID]:
        """A list of all OTUs tracked in the index."""
        return {UUID(row[0]) for row in self.con.execute("SELECT id FROM otus")}

    def add_event_id(self, event_id: int, otu_id: UUID) -> None:
        """Add a new event ID and associated it with a given OTU ID.

        This allows fast lookup of event IDs for a given OTU ID.

        :param event_id: the event ID to add
        :param otu_id: the OTU ID to add the event ID to
        """
        self.con.execute(
            """
            INSERT INTO events (event_id, otu_id)
            VALUES (?, ?)
            """,
            (
                event_id,
                str(otu_id),
            ),
        )

    def get_event_ids_by_otu_id(self, otu_id: UUID) -> EventIndexItem | None:
        """Get a list of event IDs for an OTU."""
        res = self.con.execute(
            """
            SELECT event_id
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (str(otu_id),),
        )

        event_ids = [row[0] for row in res]

        if event_ids:
            return EventIndexItem(
                event_ids[-1],
                event_ids,
                otu_id,
            )

        return None

    def get_id_by_legacy_id(self, legacy_id: str) -> UUID | None:
        """Get an OTU ID by its legacy ID."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE legacy_id = ?",
            (legacy_id,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def get_id_by_name(self, name: str) -> UUID | None:
        """Get an OTU ID by its name."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE name = ?",
            (name,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def get_id_by_taxid(self, taxid: int) -> UUID | None:
        """Get an OTU ID by its taxonomy ID."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE taxid = ?",
            (taxid,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def delete_otu(self, otu_id: UUID) -> None:
        """Remove an OTU from the index.

        This happens rarely, so we just rewrite the whole index.

        :param otu_id: the ID of the OTU to remove.
        """
        self.con.execute(
            "DELETE FROM events WHERE otu_id = ?",
            (str(otu_id),),
        )

        self.con.execute(
            "DELETE FROM otus WHERE id = ?",
            (str(otu_id),),
        )

        self.con.execute(
            "DELETE FROM sequences WHERE otu_id = ?",
            (str(otu_id),),
        )

        self.con.commit()

    def iter_minimal_otus(self) -> Iterator[OTUMinimal]:
        """Iterate over minimal representations of all OTUs in the index."""
        cursor = self.con.execute(
            "SELECT acronym, id, legacy_id, name, taxid FROM otus ORDER BY name",
        )

        for row in cursor:
            yield OTUMinimal(
                acronym=row[0],
                id=UUID(row[1]),
                legacy_id=row[2],
                name=row[3],
                taxid=row[4],
            )

    def load_snapshot(self, otu_id: UUID) -> Snapshot | None:
        """Load an OTU snapshot."""
        cursor = self.con.execute(
            "SELECT at_event, otu FROM otus WHERE id = ?",
            (str(otu_id),),
        )

        result = cursor.fetchone()

        if not result:
            return None

        at_event, otu_json = result

        otu = orjson.loads(otu_json)

        sequence_ids = [
            str(sequence["id"])
            for isolate in otu["isolates"]
            for sequence in isolate["sequences"]
        ]

        # Fetch all sequences in a single query
        placeholders = ",".join("?" for _ in sequence_ids)

        sequence_map = dict(
            self.con.execute(
                f"SELECT id, sequence FROM sequences WHERE id IN ({placeholders})",
                sequence_ids,
            ).fetchall(),
        )

        # Update the data structure and check for missing sequences
        missing_sequences = []

        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                sequence_id = str(sequence["id"])

                if sequence_id in sequence_map:
                    sequence["sequence"] = sequence_map[sequence_id]
                else:
                    missing_sequences.append(sequence_id)

        # Raise an error if any sequences are missing
        if missing_sequences:
            raise ValueError(f"Sequences not found: {', '.join(missing_sequences)}")

        return Snapshot(
            at_event=at_event,
            otu=RepoOTU.model_validate(otu),
        )

    def upsert_otu(self, otu: RepoOTU, at_event: int) -> None:
        """Update the index based on a list of OTUs.

        OTUs that aren't already in the index will be added, and those that are will be
        updated.

        Only sequences that have changed will be updated.

        """
        sequence_ids = {
            str(seq.id) for isolate in otu.isolates for seq in isolate.sequences
        }

        otu_dict = otu.model_dump(mode="json")

        placeholders = ",".join("?" for _ in sequence_ids)

        # Delete any sequences that are no longer in the OTU.
        self.con.execute(
            f"""
            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({ placeholders });
            """,
            (str(otu.id), *sequence_ids),
        )

        batch = []

        # Insert or update the sequences.
        for isolate in otu_dict["isolates"]:
            for sequence in isolate["sequences"]:
                crc = _calculate_crc32(sequence["sequence"])
                batch.append(
                    (
                        str(sequence["id"]),
                        crc,
                        str(otu.id),
                        sequence["sequence"],
                    ),
                )

        # Perform batch insert
        self.con.executemany(
            """
            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
            VALUES (?, ?, ?, ?)
            """,
            batch,
        )

        # Update only if CRC has changed
        self.con.executemany(
            """
            UPDATE sequences
            SET crc = ?, otu_id = ?, sequence = ?
            WHERE id = ? AND crc != ?
            """,
            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
        )

        self.con.execute(
            """
            INSERT OR REPLACE INTO otus (id, acronym, at_event, legacy_id, name, otu, taxid)
            VALUES (?, ?,?, ?, ?, ?, ?)
            """,
            (
                str(otu.id),
                otu.acronym,
                at_event,
                otu.legacy_id,
                otu.name,
                orjson.dumps(otu_dict),
                otu.taxid,
            ),
        )

        self.con.commit()


def _calculate_crc32(sequence: str) -> str:
    """Calculate the CRC32 checksum of a sequence.

    :param sequence: the sequence to calculate the checksum of.
    :return: the CRC32 checksum as a hexadecimal string.

    """
    crc = binascii.crc32(sequence.encode("utf-8"))

    # Convert CRC as a hexadecimal string.
    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)


def _default_json(obj: Any) -> list:
    """Cast sets to lists.

    This function is used as the default argument to `orjson.dumps` to handle sets.

    :param obj: the object to serialize
    :return: the serialized object
    """
    if isinstance(obj, set):
        return list(obj)

    raise TypeError(f"Object of type {obj.__class__.__name__} is not JSON serializable")
