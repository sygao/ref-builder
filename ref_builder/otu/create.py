import sys

import structlog

from ref_builder.ncbi.client import NCBIClient
from ref_builder.otu.utils import (
    create_isolate_plan_from_records,
    group_genbank_records_by_isolate,
    get_molecule_from_records,
    parse_refseq_comment,
)
from ref_builder.repo import Repo
from ref_builder.resources import RepoOTU

logger = structlog.get_logger("otu.create")


def create_otu(
    repo: Repo,
    taxid: int,
    accessions: list[str],
    acronym: str,
    ignore_cache: bool = False,
) -> RepoOTU | None:
    """Create a new OTU by taxonomy ID and autogenerate a schema.

    Uses the provided accessions to generate a schema and add a first isolate.

    :param repo: the repository to add the OTU to.
    :param taxid: the taxonomy ID to use.
    :param accessions: a list of accessions to use for the schema.
    :param acronym: an alternative name to use during searches.
    :param ignore_cache: whether to ignore the cache.

    """
    otu_logger = logger.bind(taxid=taxid)

    client = NCBIClient(ignore_cache)

    if repo.get_otu_id_by_taxid(taxid):
        raise ValueError(
            f"Taxonomy ID {taxid} has already been added to this reference.",
        )

    taxonomy = client.fetch_taxonomy_record(taxid)

    if taxonomy is None:
        otu_logger.fatal(f"Could not retrieve {taxid} from NCBI Taxonomy")
        return None

    if not acronym and taxonomy.other_names.acronym:
        acronym = taxonomy.other_names.acronym[0]

    records = client.fetch_genbank_records(accessions)

    if len(records) != len(accessions):
        otu_logger.fatal("Could not retrieve all requested accessions.")
        return None

    binned_records = group_genbank_records_by_isolate(records)

    if len(binned_records) > 1:
        otu_logger.fatal(
            "More than one isolate found. Cannot create schema automatically.",
        )
        return None

    molecule = get_molecule_from_records(records)

    plan = create_isolate_plan_from_records(
        records,
        length_tolerance=repo.settings.default_segment_length_tolerance,
    )

    if plan is None:
        otu_logger.fatal("Could not create plan from records.")
        return None

    try:
        otu = repo.create_otu(
            acronym=acronym,
            legacy_id=None,
            molecule=molecule,
            name=taxonomy.name,
            plan=plan,
            taxid=taxid,
        )
    except ValueError as e:
        otu_logger.fatal(e)
        sys.exit(1)

    isolate = repo.create_isolate(
        otu_id=otu.id,
        legacy_id=None,
        name=next(iter(binned_records.keys())) if binned_records else None,
    )

    otu.add_isolate(isolate)
    otu.repr_isolate = repo.set_repr_isolate(otu_id=otu.id, isolate_id=isolate.id)

    for record in records:
        sequence = repo.create_sequence(
            otu_id=otu.id,
            accession=record.accession_version,
            definition=record.definition,
            legacy_id=None,
            segment=record.source.segment,
            sequence=record.sequence,
        )

        repo.link_sequence(otu.id, isolate.id, sequence.id)

        if record.refseq:
            _, old_accession = parse_refseq_comment(record.comment)
            repo.exclude_accession(
                otu.id,
                old_accession,
            )

    return repo.get_otu(otu.id)
