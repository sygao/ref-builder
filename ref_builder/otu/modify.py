from pydantic import ValidationError

from ref_builder.ncbi.client import NCBIClient
from ref_builder.otu.update import logger
from ref_builder.otu.utils import create_segments_from_records
from ref_builder.plan import (
    MonopartitePlan,
    MultipartitePlan,
    SegmentRule,
    SegmentName,
    Segment,
)
from ref_builder.repo import Repo
from ref_builder.resources import RepoOTU


def set_isolate_plan(
    repo: Repo,
    otu: RepoOTU,
    plan: MonopartitePlan | MultipartitePlan,
) -> MonopartitePlan | MultipartitePlan | None:
    """Sets an OTU's isolate plan to a new plan."""
    otu_logger = logger.bind(name=otu.name, taxid=otu.taxid, plan=plan.model_dump())

    try:
        repo.set_isolate_plan(otu.id, plan)
    except ValueError as e:
        otu_logger.error(e)
        return None

    return repo.get_otu(otu.id).plan


def set_plan_length_tolerances(
    repo: Repo,
    otu: RepoOTU,
    tolerance: float,
) -> MonopartitePlan | MultipartitePlan | None:
    """Sets a plan's length tolerances to a new float value."""
    if otu.plan.plan_type == "monopartite":
        try:
            repo.set_isolate_plan(
                otu.id,
                MonopartitePlan.new(
                    length=otu.plan.length,
                    length_tolerance=tolerance,
                ),
            )
        except (ValueError, ValidationError) as e:
            logger.error(
                e,
                name=otu.name,
                taxid=otu.taxid,
                tolerance=otu.plan.length_tolerance,
                new_tolerance=tolerance,
            )
            return None

    return repo.get_otu(otu.id).plan


def add_segments_to_plan(
    repo: Repo,
    otu: RepoOTU,
    rule: SegmentRule,
    accessions: list[str],
    ignore_cache: bool = False,
) -> MultipartitePlan | None:
    """Add new segments to a multipartite plan."""
    expand_logger = logger.bind(
        name=otu.name, taxid=otu.taxid, accessions=accessions, rule=rule
    )

    sequence_length_tolerance = repo.settings.default_segment_length_tolerance

    if type(otu.plan) is MonopartitePlan:
        raise ValueError("Cannot add segments to a monopartite plan.")

    client = NCBIClient(ignore_cache)

    expand_logger.info(
        f"Adding {len(accessions)} sequences to plan as {rule} segments",
        current_plan=otu.plan.model_dump(),
    )

    records = client.fetch_genbank_records(accessions)
    if not records:
        expand_logger.warning("Could not fetch records.")
        return None

    new_segments = create_segments_from_records(
        records,
        rule,
        length_tolerance=sequence_length_tolerance,
    )
    if not new_segments:
        expand_logger.warning("No segments can be added.")
        return None

    new_plan = MultipartitePlan.new(
        segments=otu.plan.segments + new_segments,
    )

    return set_isolate_plan(repo, otu, new_plan)


def resize_monopartite_plan(
    repo: Repo,
    otu: RepoOTU,
    name: SegmentName,
    rule: SegmentRule,
    accessions: list[str],
    # sequence_length_tolerance: float | None = None,
    ignore_cache: bool = False,
) -> MultipartitePlan | None:
    """Convert a monopartite plan to a multipartite plan and add segments."""
    expand_logger = logger.bind(
        name=otu.name, taxid=otu.taxid, accessions=accessions, rule=rule
    )

    if type(otu.plan) is MultipartitePlan:
        expand_logger.warning("OTU plan is already a multipartite plan.")
        return None

    sequence_length_tolerance = repo.settings.default_segment_length_tolerance

    client = NCBIClient(ignore_cache)

    records = client.fetch_genbank_records(accessions)
    if not records:
        expand_logger.warning("Could not fetch records.")
        return None

    new_segments = create_segments_from_records(
        records,
        rule,
        length_tolerance=sequence_length_tolerance,
    )
    if not new_segments:
        expand_logger.warning("No segments can be added.")
        return None

    new_plan = MultipartitePlan.new(
        segments=[
            Segment.new(
                length=otu.plan.length,
                length_tolerance=sequence_length_tolerance,
                name=name,
                required=SegmentRule.REQUIRED,
            )
        ]
        + new_segments,
    )

    return set_isolate_plan(repo, otu, new_plan)
