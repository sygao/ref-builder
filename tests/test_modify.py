import subprocess
from uuid import UUID

import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.repo import Repo
from ref_builder.resources import RepoSequence
from ref_builder.otu.create import create_otu
from ref_builder.otu.modify import (
    add_segments_to_plan,
    resize_monopartite_plan,
    set_isolate_plan,
    set_plan_length_tolerances,
)
from ref_builder.otu.update import (
    add_genbank_isolate,
    promote_otu_accessions,
    replace_sequence_in_otu,
    set_representative_isolate,
    delete_isolate_from_otu,
)
from ref_builder.utils import IsolateName, IsolateNameType
from ref_builder.plan import (
    MonopartitePlan,
    MultipartitePlan,
    SegmentName,
    SegmentRule,
    Segment,
)


def test_update_representative_isolate(scratch_repo: Repo):
    """Test representative isolate replacement."""
    taxid = 345184

    otu_before = scratch_repo.get_otu_by_taxid(taxid)

    repr_isolate_after = None
    for isolate_id in otu_before.isolate_ids:
        if isolate_id != otu_before.repr_isolate:
            repr_isolate_after = isolate_id
            break

    set_representative_isolate(scratch_repo, otu_before, repr_isolate_after)

    otu_after = scratch_repo.get_otu_by_taxid(taxid)

    assert otu_after.repr_isolate != otu_before.repr_isolate

    assert otu_after.repr_isolate == repr_isolate_after


class TestSetIsolatePlan:
    def test_set_isolate_plan(self, scratch_repo: Repo):
        otu_before = scratch_repo.get_otu_by_taxid(223262)

        original_plan = otu_before.plan

        assert type(original_plan) is MultipartitePlan

        new_isolate_plan = MultipartitePlan.new(segments=original_plan.segments)

        new_isolate_plan.segments.append(
            Segment.new(
                length=2000,
                length_tolerance=scratch_repo.settings.default_segment_length_tolerance,
                name=SegmentName(prefix="DNA", key="C"),
                required=SegmentRule.RECOMMENDED,
            ),
        )

        new_isolate_plan.segments.append(
            Segment.new(
                length=1000,
                length_tolerance=scratch_repo.settings.default_segment_length_tolerance,
                name=SegmentName(prefix="DNA", key="Z"),
                required=SegmentRule.OPTIONAL,
            ),
        )

        set_isolate_plan(scratch_repo, otu_before, new_isolate_plan)

        assert type(new_isolate_plan) is MultipartitePlan

        otu_after = scratch_repo.get_otu(otu_before.id)

        assert len(otu_after.plan.segments) == len(otu_before.plan.segments) + 2

        assert otu_after.plan == new_isolate_plan

    def test_add_segments_to_plan(
        self,
        precached_repo: Repo,
        snapshot: SnapshotAssertion,
    ):
        otu_before = create_otu(
            precached_repo,
            2164102,
            ["MF062136", "MF062137"],
            acronym="",
        )

        original_plan = otu_before.plan

        assert type(original_plan) is MultipartitePlan

        expanded_plan = add_segments_to_plan(
            precached_repo,
            otu_before,
            rule=SegmentRule.OPTIONAL,
            accessions=["MF062138"],
        )

        assert len(expanded_plan.segments) == len(original_plan.segments) + 1

        otu_after = precached_repo.get_otu(otu_before.id)

        assert otu_after.plan != otu_before.plan

        assert otu_after.plan.model_dump() == snapshot(exclude=props("id"))

    def test_resize_monopartite_plan(
        self,
        precached_repo: Repo,
        snapshot: SnapshotAssertion,
    ):
        otu_before = create_otu(
            precached_repo,
            2164102,
            ["MF062136"],
            acronym="",
        )

        assert type(otu_before.plan) is MonopartitePlan

        resize_monopartite_plan(
            precached_repo,
            otu_before,
            name=SegmentName(prefix="RNA", key="L"),
            rule=SegmentRule.RECOMMENDED,
            accessions=["MF062137", "MF062138"],
        )

        otu_after = precached_repo.get_otu(otu_before.id)

        assert type(otu_after.plan) is MultipartitePlan

        assert otu_after.plan.required_segments[0].length == otu_before.plan.length

        assert otu_after.plan.model_dump() == snapshot(exclude=props("id"))

    def test_extend_plan_monopartite_fail(
        self,
        scratch_repo: Repo,
    ):
        """Test that add_segments_to_plan() fails out
        when the original plan is a MonopartitePlan.
        """
        otu_before = scratch_repo.get_otu_by_taxid(96892)

        assert type(otu_before.plan) is MonopartitePlan

        with pytest.raises(ValueError):
            add_segments_to_plan(
                scratch_repo,
                otu_before,
                rule=SegmentRule.OPTIONAL,
                accessions=["NC_010620"],
            )

    @pytest.mark.parametrize("tolerance", [0.05, 0.5, 1.0])
    def test_set_length_tolerances_ok(self, scratch_repo: Repo, tolerance: float):
        """Check that plan length tolerances can be modified by function."""
        otu_before = scratch_repo.get_otu_by_taxid(96892)

        assert (
            otu_before.plan.length_tolerance
            == scratch_repo.settings.default_segment_length_tolerance
        )

        set_plan_length_tolerances(scratch_repo, otu_before, tolerance)

        otu_after = scratch_repo.get_otu(otu_before.id)

        assert otu_after.plan.length_tolerance == tolerance

    @pytest.mark.parametrize("bad_tolerance", [-1.0, 1.1, 100.0])
    def test_set_length_tolerances_fail(self, scratch_repo: Repo, bad_tolerance: float):
        """Check that plan length tolerances cannot be set to an invalid float value."""
        otu_before = scratch_repo.get_otu_by_taxid(96892)

        assert (
            otu_before.plan.length_tolerance
            == scratch_repo.settings.default_segment_length_tolerance
        )

        set_plan_length_tolerances(scratch_repo, otu_before, bad_tolerance)

        otu_after = scratch_repo.get_otu(otu_before.id)

        assert otu_after.plan.length_tolerance == otu_before.plan.length_tolerance


class TestUpdateRepresentativeIsolateCommand:
    def test_isolate_id_ok(self, scratch_repo: Repo):
        taxid = 345184

        otu_before = scratch_repo.get_otu_by_taxid(taxid)

        repr_isolate_after = None

        for isolate_id in otu_before.isolate_ids:
            if isolate_id != otu_before.repr_isolate:
                repr_isolate_after = isolate_id
                break

        subprocess.run(
            [
                "ref-builder",
                "otu",
                "update",
            ]
            + ["--path", str(scratch_repo.path)]
            + [str(taxid)]
            + ["default"]
            + [str(repr_isolate_after)],
            check=False,
        )

        scratch_repo = Repo(scratch_repo.path)

        otu_after = scratch_repo.get_otu_by_taxid(taxid)

        assert otu_after.repr_isolate != otu_before.repr_isolate

        assert otu_after.repr_isolate == repr_isolate_after

    def test_isolate_name_ok(self, scratch_repo: Repo):
        taxid = 1169032

        otu_before = scratch_repo.get_otu_by_taxid(taxid)

        repr_isolate_name_after = None

        for isolate_id in otu_before.isolate_ids:
            if isolate_id != otu_before.repr_isolate:
                repr_isolate_name_after = otu_before.get_isolate(isolate_id).name
                break

        subprocess.run(
            [
                "ref-builder",
                "otu",
                "update",
            ]
            + ["--path", str(scratch_repo.path)]
            + [str(taxid)]
            + ["default"]
            + [str(repr_isolate_name_after)],
            check=False,
        )

        scratch_repo = Repo(scratch_repo.path)

        otu_after = scratch_repo.get_otu_by_taxid(taxid)

        assert otu_after.repr_isolate != otu_before.repr_isolate

        assert otu_after.repr_isolate == otu_before.get_isolate_id_by_name(
            repr_isolate_name_after
        )


class TestDeleteIsolate:
    def test_ok(self, scratch_repo):
        """Test that a given isolate can be deleted from the OTU."""
        taxid = 1169032

        otu_before = scratch_repo.get_otu_by_taxid(taxid)

        isolate_id = otu_before.get_isolate_id_by_name(
            IsolateName(type=IsolateNameType.ISOLATE, value="WMoV-6.3"),
        )

        assert type(isolate_id) is UUID

        delete_isolate_from_otu(scratch_repo, otu_before, isolate_id)

        otu_after = scratch_repo.get_otu_by_taxid(taxid)

        assert isolate_id not in otu_after.isolate_ids

        assert otu_after.get_isolate(isolate_id) is None

        assert otu_before.get_isolate(isolate_id).accessions not in otu_after.accessions

        assert len(otu_after.isolate_ids) == len(otu_before.isolate_ids) - 1


class TestReplaceSequence:
    def test_ok(self, precached_repo):
        """Test sequence replacement and deletion."""
        otu_before = create_otu(
            precached_repo,
            1169032,
            ["MK431779"],
            acronym="",
        )

        isolate_id, old_sequence_id = (
            otu_before.get_sequence_id_hierarchy_from_accession(
                "MK431779",
            )
        )

        assert type(old_sequence_id) is UUID

        sequence = replace_sequence_in_otu(
            repo=precached_repo,
            otu=otu_before,
            new_accession="NC_003355",
            replaced_accession="MK431779",
        )

        assert type(sequence) is RepoSequence

        otu_after = precached_repo.get_otu_by_taxid(1169032)

        assert (
            otu_after.accessions
            == otu_after.get_isolate(isolate_id).accessions
            == {"NC_003355"}
        )


@pytest.mark.ncbi()
class TestPromoteAccessions:
    def test_ok(self, empty_repo: Repo):
        """Test that RefSeq accessions can be promoted automatically."""
        otu = create_otu(
            empty_repo, 2164102, ["MF062136", "MF062137", "MF062138"], acronym=""
        )
        isolate = add_genbank_isolate(
            empty_repo, otu, ["MF062125", "MF062126", "MF062127"]
        )

        otu_before = empty_repo.get_otu(otu.id)

        assert otu_before.accessions == {
            "MF062125",
            "MF062126",
            "MF062127",
            "MF062136",
            "MF062137",
            "MF062138",
        }

        assert otu_before.get_isolate(isolate.id).accessions == {
            "MF062125",
            "MF062126",
            "MF062127",
        }

        promoted_accessions = promote_otu_accessions(empty_repo, otu_before)

        assert promoted_accessions == {"NC_055390", "NC_055391", "NC_055392"}

        otu_after = empty_repo.get_otu(otu.id)

        assert otu_after.isolate_ids == otu_before.isolate_ids

        assert otu_after.accessions == {
            "NC_055390",
            "NC_055391",
            "NC_055392",
            "MF062136",
            "MF062137",
            "MF062138",
        }

        assert otu_after.get_isolate(isolate.id).accessions == {
            "NC_055390",
            "NC_055391",
            "NC_055392",
        }

        assert otu_after.excluded_accessions == {"MF062125", "MF062126", "MF062127"}

    def test_command_ok(self, empty_repo: Repo):
        otu = create_otu(
            empty_repo, 2164102, ["MF062125", "MF062126", "MF062127"], acronym=""
        )

        otu_before = empty_repo.get_otu(otu.id)

        assert otu_before.accessions == {"MF062125", "MF062126", "MF062127"}

        subprocess.run(
            [
                "ref-builder",
                "otu",
                "update",
            ]
            + ["--path", str(empty_repo.path)]
            + [str(2164102)]
            + ["promote"],
            check=False,
        )

        repo_after = Repo(empty_repo.path)

        otu_after = repo_after.get_otu(otu.id)

        assert otu_after.repr_isolate == otu_before.repr_isolate

        assert otu_after.accessions == {"NC_055390", "NC_055391", "NC_055392"}
