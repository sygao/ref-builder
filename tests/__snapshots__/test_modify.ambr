# serializer version: 1
# name: TestSetIsolatePlan.test_add_segments_to_plan
  dict({
    'plan_type': 'multipartite',
    'segments': list([
      dict({
        'length': 7219,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'L',
          'prefix': 'RNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1613,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'M',
          'prefix': 'RNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1291,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'S',
          'prefix': 'RNA',
        }),
        'required': 'optional',
      }),
    ]),
  })
# ---
# name: TestSetIsolatePlan.test_resize_monopartite_plan
  dict({
    'plan_type': 'multipartite',
    'segments': list([
      dict({
        'length': 7219,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'L',
          'prefix': 'RNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1613,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'M',
          'prefix': 'RNA',
        }),
        'required': 'recommended',
      }),
      dict({
        'length': 1291,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'S',
          'prefix': 'RNA',
        }),
        'required': 'recommended',
      }),
    ]),
  })
# ---
